### PR TITLE
Lookup datums by inputs to address potential exploit and general cleanup

### PR DESCRIPTION
**Pull request by jfischoff**
Instead of folding over all datums, we map over the script inputs and lookup each associated datum. This allows use to collect duplicate datums, which occur in some uncommon situations.

Additionally, we remove used code and make some slightly reorganizations for improved readability.

The tests were also updated, and a test was added for using duplicate datums.

### CODE REVIEW

File: src/Canonical/JpgStore/BulkPurchase.hs [hunk 1]
```diff
@@ -225,36 +205,45 @@ swapValidator _ r SwapScriptContext{sScriptContextTxInfo = SwapTxInfo{..}, sScri
...
     swaps :: [Swap]
-    swaps = map (\(_, d) -> convertDatum d) sTxInfoData
+    swaps = mapMaybe
+      (\i -> fmap (\d -> convertDatum (lookupDatum d)) (sTxOutDatumHash (sTxInInfoResolved i))) scriptInputs
 
     outputsAreValid :: Map PubKeyHash Value -> Bool
     outputsAreValid = validateOutputConstraints sTxInfoOutputs
 
-    foldSwaps :: (Swap -> Map PubKeyHash Value -> Map PubKeyHash Value) -> Map PubKeyHash Value -> Map PubKeyHash Value
-    foldSwaps f init = foldr f init swaps
```
Comment (jfischoff) [inline src/Canonical/JpgStore/BulkPurchase.hs:235]: We inlined this function because we only use it once

File: src/Canonical/JpgStore/BulkPurchase.hs [hunk 2]
```diff
@@ -142,29 +142,13 @@ isScriptThisInput vh txIn = case sAddressCredential (sTxOutAddress  (sTxInInfoRe
...
-        { sTxOutAddress = SwapAddress
-          { sAddressCredential = ScriptCredential vh
-          }
-        }
-      } : xs ->  if vh == thisValidator then
-              go xs
-            else
-              TRACE_IF_FALSE("Bad validator input", "100", False)
-    _ : xs -> go xs
-
```
Comment (jfischoff) [inline src/Canonical/JpgStore/BulkPurchase.hs:160]: Unused

File: src/Canonical/JpgStore/BulkPurchase.hs [hunk 3]
```diff
@@ -225,36 +205,45 @@ swapValidator _ r SwapScriptContext{sScriptContextTxInfo = SwapTxInfo{..}, sScri
...
-  in if sTxInInfoOutRef (head (filter (isScriptThisInput thisValidator) sTxInfoInputs)) /= thisOutRef then True else
+  in if sTxInInfoOutRef (head scriptInputs) /= thisOutRef then True else
     case r of
       Cancel ->
         let
+          signerIsOwner :: Swap -> Bool
           signerIsOwner Swap{sOwner} = singleSigner == sOwner
-        in
-          TRACE_IF_FALSE("signer is not the owner", "4", (all signerIsOwner swaps))
+        in TRACE_IF_FALSE("signer is not the owner", "4", (all signerIsOwner swaps))
```
Comment (jfischoff) [inline src/Canonical/JpgStore/BulkPurchase.hs:232]: Small readability changes
